### PR TITLE
skip transpose in array.gufunc when possible

### PR DESCRIPTION
**Pull request by rabernat**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

We are using and loving `gufunc` in a [new package](https://github.com/xgcm/xcape) which wraps a custom fortran code. Thanks for this awesome feature! (I know it's been here for a while, but I just discovered it.)

In our application, the function we want to apply is already behaves as a gufunc at the numpy level--it operates along one core dimension (axis -1) and vectorizes (within fortran) over all preceding dimensions. So we don't necessarily need all of the transposing / reshaping that gufunc provides, just the ability to parallelize an existing gufunc.

On the other end of the performance spectrum, I'm also interested in minimizing the number of tasks in the dask graph. This PR adds a small check that eliminates an unnecessary transpose step on the inputs of gufunc in the (likely quite common) case that the inputs don't have to be transposed.

I wasn't sure how to test for this, since it is a purely internal optimization.

### CODE REVIEW

File: dask/array/gufunc.py [hunk 1]
```diff
@@ -351,8 +351,11 @@ def apply_gufunc(func, signature, *args, **kwargs):
         shape = arg.shape
         iax = tuple(a if a < 0 else a - len(shape) for a in iax)
         tidc = tuple(i for i in range(-len(shape) + 0, 0) if i not in iax) + iax
-
-        transposed_arg = arg.transpose(tidc)
+        if tidc == tuple(range(-arg.ndim, 0)):
+            # transpose not necessary
```
Comment (TomAugspurger) [inline dask/array/gufunc.py]: For my understanding, this means that `arg == arg.transpose(tidc)`?

Comment (rabernat) [reply dask/array/gufunc.py]: correct

**Comment by rabernat** [comment]
> We could look at the number of tasks. But I'm not sure how we could write a robust test. I think we're fine without tests here.

FWIW, I did some ad-hoc testing with my data and did confirm it reduced the number of tasks as expected.

File: dask/array/gufunc.py (hunk 1 above)
Comment (mrocklin) [reply dask/array/gufunc.py]: We might also make this fix in the transpose function itself.  That would also be easier to test.

```python
assert x.transpose((1, 2, 3)) is x
```

**Comment by mrocklin** [comment]
+1 from me.  Thanks @rabernat !

**Comment by jrbourbeau** [comment]
Thanks @rabernat, will merge on CI passing

**Pull request merged**